jtlibrary/python/jtmodules/src/jtmodules/generate_volume_image.py

```diff
@@ -66,80 +134,218 @@ def fit_plane(points):
-    else:
-        image_binned = image
-
-    x, y, z = [], [], []
-    for ix in range(image.shape[0]):
-        for iy in range(image.shape[1]):
-            if mask[ix, iy] > 0:
-                max_pos = np.argmax(
-                    image_binned[int(ix / bin_size), int(iy / bin_size), :]
-                )
-                height = int(
-                    max_pos - plane(ix, iy, surface_plane_params)
-                )
-                if height > 0:
-                    x.append(ix)
-                    y.append(iy)
-                    z.append(height)
-
-    coords = collections.namedtuple("coords", ["x", "y", "z"])
-    return coords(x=x, y=y, z=z)
-
-
 def interpolate_surface(coords, output_shape, method='linear'):
     '''Given a set of coordinates (not necessarily on a grid), an
     interpolation is returned as a numpy array'''
     from scipy.interpolate import griddata
 
-    xy = np.column_stack((coords.x, coords.y))
+    coords = coords.astype(np.float32)
+
+    xy = np.column_stack((coords[:, 0], coords[:, 1]))
     xv, yv = np.meshgrid(
         range(output_shape[0]),
         range(output_shape[1])
     )
     if method == 'nearest':
         interpolate = griddata(
-            xy, np.array(coords.z), (xv, yv), method='nearest', rescale=False
+            xy, np.array(coords[:, 2]), (xv, yv), method='nearest', rescale=False
         )
     elif method == 'cubic':
         interpolate = griddata(
-            xy, np.array(coords.z), (xv, yv), method='cubic', rescale=False
+            xy, np.array(coords[:, 2]), (xv, yv), method='cubic', rescale=False
         )
     elif method == 'linear':
         interpolate = griddata(
-            xy, np.array(coords.z), (xv, yv), method='linear', rescale=False
+            xy, np.array(coords[:, 2]), (xv, yv), method='linear', rescale=False
         )
 
     return interpolate.T
 
 
-def main(image, mask, threshold=150, bead_size=2, superpixel_size=4,
-         close_surface=False, close_disc_size=8, plot=False):
+def localise_bead_maxima_3D(image, labeled_beads, minimum_bead_intensity):
+    '''Given a 3D image of beads, and a 3D segmentation image,
+    bead maxima are returned as a numpy array'''
+
+    logger.debug('removing beads with intensity < %d', minimum_bead_intensity)
+    bead_coords = []
+    bboxes = mh.labeled.bbox(labeled_beads)
+    for bead in range(np.max(labeled_beads)):
+        x_min, x_max, y_min, y_max, z_min, z_max = bboxes[bead]
+        label_image = labeled_beads[x_min:x_max, y_min:y_max, z_min:z_max]
+        bounded = image[x_min:x_max, y_min:y_max, z_min:z_max]
+        bounded = np.copy(bounded)
+        bounded[label_image != bead] = 0
+        local_coords = np.unravel_index(
+            bounded.argmax(),
+            bounded.shape
+        )
+        bbox_min = (x_min, y_min, z_min)
+        centre = tuple(a + b for a, b in zip(bbox_min, local_coords))
+
+        if (image[centre[0], centre[1], centre[2]] > minimum_bead_intensity):
```

Comment:
Is `centre` a tuple of 3 elements? Then you can let Python expand it:
```suggestion
        if image[centre] > minimum_bead_intensity:
```